pilot/pkg/serviceregistry/serviceentry/store.go

```diff
@@ -63,8 +63,16 @@ func (s *serviceInstancesStore) deleteInstances(key configKey, instances []*mode
 		ikey := makeInstanceKey(i)
 		hostPort := hostPort{ikey.hostname.String(), i.ServicePort.Port}
 		s.instancesByHostAndPort.Delete(hostPort)
-		delete(s.instances[makeInstanceKey(i)], key)
+		oldInstances := s.instances[ikey][key]
+		delete(s.instances[ikey], key)
+		if len(s.instances[ikey]) == 0 {
+			delete(s.instances, ikey)
+		}
 		delete(s.ip2instance, i.Endpoint.Address)
+		// Cleanup stale IPs, if the IPs changed
+		for _, oi := range oldInstances {
```

Comment:
I think it is not the right place to remove old instances. From function `deleteInstances`, we meant to remove the specified instances.
haven't run the test without this, but i would think if this is necessary, we should put it in the caller of this function, it should be in se/wle handler

Reply:
After looking into this a bit more, I think it could be changed to take in instanceKeys to be more clear. However, this is a bit less efficient so I instead renamed the function and added documentation. LMK if you have other suggestions